Sources/AsyncHTTPClient/HTTPHandler.swift

```diff
@@ -473,12 +473,14 @@ extension HTTPClient {
 
         /// Cancels the request execution.
         public func cancel() {
-            self.lock.withLock {
+            let channel: Channel? = self.lock.withLock {
                 if !cancelled {
                     cancelled = true
-                    channel?.pipeline.fireUserInboundEventTriggered(TaskCancelEvent())
+                    return self.channel
                 }
+                return nil
             }
+            channel?.triggerUserOutboundEvent(TaskCancelEvent(), promise: nil)
```

Comment:
@artemredkin I'm pretty sure this doesn't actually work (which proves there's no test) because whatever ChannelHandler is supposed to catch this event needs to change from inbound to outbound user event.

Reply:
we need to add a test that verifies cancellation works.

Reply:
fixed! and tests pass now